### PR TITLE
Refactor the services package into 3 smaller packages: file, program and fileerror.

### LINKED ISSUE #20: Package pkg/services should be split in smaller packages, one for each service

### PR DESCRIPTION
**Pull request by joaogsousa**
Fixes #20

### CODE REVIEW

**Comment by matheuscscp** [comment]
Thanks for the contribution! I'll review it. Please add `Fixes #20` to the description, so we close the issue automatically.

File: pkg/services/fileerror/file_error.go [hunk 1]
```diff
@@ -1,4 +1,4 @@
-package services
+package fileerror
```
Comment (matheuscscp) [inline pkg/services/fileerror/file_error.go]: I think it's better if we merge `pkg/services/fileerror` into `pkg/services/file`. What you think?

File: test/mocks/gen/pkg/services/file/file.go [hunk 2]
```diff
@@ -0,0 +1,425 @@
+// Code generated by MockGen. DO NOT EDIT.
```
Comment (matheuscscp) [inline test/mocks/gen/pkg/services/file/file.go]: We need a clever way to remove old mock files when we make a rename like this. `test/mocks/gen/pkg/services/file.go` and `test/mocks/gen/pkg/services/program.go` should be removed... We need to think about something to improve the `Makefile` for that... Would you please remove these two file manually until we have a better `Makefile`?

File: cmd/judge.go [hunk 3]
```diff
@@ -73,11 +74,11 @@ func parseExecuteFlags(cmd *cobra.Command, flags *executeFlags) (*judge.Executor
 	if err := json.Unmarshal([]byte(flags.solutionRequestHeaders), &solutionHeaders); err != nil {
 		return nil, fmt.Errorf("error unmarshaling problem solution request headers: %w", err)
 	}
-	interactorProgramService, err := services.NewProgramService(flags.interactorProgramService, nil)
+	interactorProgramService, err := program.NewProgramService(flags.interactorProgramService, nil)
```
Comment (matheuscscp) [inline cmd/judge.go]: As we spoke, we're changing `NewProgramService` and `NewFileService` to `NewService` in both packages.

File: cmd/judge.go [hunk 4]
```diff
@@ -122,7 +123,7 @@ func parseExecuteFlags(cmd *cobra.Command, flags *executeFlags) (*judge.Executor
 
 // bindExecuteFlags binds judge execute command flags.
 func bindExecuteFlags(cmd *cobra.Command, flags *executeFlags) {
-	availableProgramServices := strings.Join(services.GetProgramServices(), ", ")
+	availableProgramServices := strings.Join(program.GetProgramServices(), ", ")
```
Comment (matheuscscp) [inline cmd/judge.go]: `GetProgramServices` -> `GetServices`

File: judge/executor_test.go [hunk 5]
```diff
@@ -70,13 +71,13 @@ func TestExecute(t *testing.T) {
 			err = programFactory.Create(test.solutionProgram, "./serverFiles/solution")
 			assert.Equal(t, nil, err)
 
-			var interactorProgramService services.ProgramService
+			var interactorProgramService program.Service
 			if test.interactor == judge.CustomInteractor {
-				interactorProgramService, err = services.NewProgramService(test.interactorProgramService, nil)
+				interactorProgramService, err = program.NewProgramService(test.interactorProgramService, nil)
```
Comment (rodopoulos) [inline judge/executor_test.go]: nit: `program.NewProgramService` -> `program.NewService`

File: judge/executor_test.go [hunk 6]
```diff
@@ -70,13 +71,13 @@ func TestExecute(t *testing.T) {
...
 				assert.Equal(t, nil, err)
 			}
 
-			solutionProgramService, err := services.NewProgramService(test.solutionProgramService, nil)
+			solutionProgramService, err := program.NewProgramService(test.solutionProgramService, nil)
```
Comment (rodopoulos) [inline judge/executor_test.go]: nit: `program.NewProgramService` -> `program.NewService`

File: pkg/services/file/file.go [hunk 7]
```diff
@@ -86,9 +88,9 @@ type (
 	}
 )
 
-// NewFileService returns a new instance of the default implementation of FileService.
-// If nil is passed, the default FileService will be created with the default defaultFileServiceRuntime.
-func NewFileService(runtime defaultFileServiceRuntime) FileService {
+// NewFileService returns a new instance of the default implementation of Service.
+// If nil is passed, the default Service will be created with the default defaultFileServiceRuntime.
+func NewFileService(runtime defaultFileServiceRuntime) Service {
```
Comment (matheuscscp) [inline pkg/services/file/file.go]: We could also rename `defaultFileService` to `defaultService`, `defaultFileServiceRuntime` to `serviceRuntime` and `fileServiceDefaultRuntime` to `defaultServiceRuntime`. What you think?

File: pkg/services/program/program.go [hunk 8]
```diff
@@ -36,9 +36,9 @@ type (
 	}
 )
 
-// NewProgramService creates a ProgramService according to the given key.
-// If nil is passed, the ProgramService will be created with the default programServiceRuntime.
-func NewProgramService(programServiceKey string, runtime programServiceRuntime) (ProgramService, error) {
+// NewProgramService creates a Service according to the given key.
+// If nil is passed, the Service will be created with the default programServiceRuntime.
+func NewProgramService(programServiceKey string, runtime programServiceRuntime) (Service, error) {
```
Comment (matheuscscp) [inline pkg/services/program/program.go]: We could rename these three names like I suggested for the file package, check my previous comment.

File: pkg/services/program/program.go [hunk 9]
```diff
@@ -64,8 +64,8 @@ func GetProgramServices() []string {
 }
 
 // getProgramServices returns the available program services.
-func getProgramServices(runtime programServiceRuntime) map[string]ProgramService {
-	return map[string]ProgramService{
+func getProgramServices(runtime programServiceRuntime) map[string]Service {
```
Comment (matheuscscp) [inline pkg/services/program/program.go]: Rename to `getServices(...)`?

File: pkg/services/file/file.go (hunk 7 above)
Comment (joaogsousa) [reply pkg/services/file/file.go]: ok

**Comment by joaogsousa** [comment]
A question...
after executing make all I reveived this line on the test report:
?   	github.com/matheuscscp/fd8-judge/judge	[no test files]

Is this expected? Or did I broke something?

**Comment by matheuscscp** [comment]
> A question...
> after executing make all I reveived this line on the test report:
> ? github.com/matheuscscp/fd8-judge/judge [no test files]
> 
> Is this expected? Or did I break something?

This was only for the `unit` tests section, right? That's correct, 'cause we still don't have any unit tests for this pkg.

**Pull request merged**